## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-tls.sh | :x: [more...](#f6252ae1-2615-4aa0-85ea-b191b605c7f7) |
| mint-xl.sh | :x: [more...](#c4ffe113-340a-45e8-ba81-0f199ecacb4d) |
| mint-large-bucket.sh | :x: [more...](#7459b5c4-2dfc-4545-97fe-0fcd0b664a26) |
| mint-dist-xl.sh | :x: [more...](#3169f83b-70ed-4fb3-aa72-100321d00e18) |
<a name="7459b5c4-2dfc-4545-97fe-0fcd0b664a26"></a>
### 6419-b657ee0/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30172
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a97ab3cb2ebe:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 1 minutes and 0 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 51 seconds
(5/13) Running aws-sdk-ruby tests ... done in 15 seconds
(6/13) Running mc tests ... done in 32 seconds
(7/13) Running minio-dotnet tests ... done in 4 minutes and 39 seconds
(8/13) Running minio-go tests ... done in 1 minutes and 41 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 1 seconds
(10/13) Running minio-js tests ... done in 41 seconds
(11/13) Running minio-py tests ... FAILED in 1 hours 5 minutes and 12 seconds
{
  "function": "put_object(bucket_name, object_name, data, length, content_type, metadata)",
  "message": "HTTPConnectionPool(host='minikube', port=30172): Max retries exceeded with url: /minio-py-test-28297ee2-223d-4069-9999-546b92b04b9c/a135b348-f47c-4433-8006-a5611635bf7a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 600, in urlopen\n    chunked=chunked)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 384, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 2, in raise_from\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 380, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.5/http/client.py\", line 1197, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.5/http/client.py\", line 297, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.5/http/client.py\", line 266, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1427, in test_thread_safe\n    client.remove_object(bucket_name, object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 942, in remove_object\n    object_name=object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1752, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 322, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 638, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 398, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minikube', port=30172): Max retries exceeded with url: /minio-py-test-28297ee2-223d-4069-9999-546b92b04b9c/a135b348-f47c-4433-8006-a5611635bf7a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1800, in main\n    test_thread_safe(client, testfile, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1430, in test_thread_safe\n    raise Exception(err)\nException: HTTPConnectionPool(host='minikube', port=30172): Max retries exceeded with url: /minio-py-test-28297ee2-223d-4069-9999-546b92b04b9c/a135b348-f47c-4433-8006-a5611635bf7a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n",
  "name": "minio-py:test_thread_safe",
  "args": {
    "bucket_name": "minio-py-test-28297ee2-223d-4069-9999-546b92b04b9c",
    "object_name": "a135b348-f47c-4433-8006-a5611635bf7a"
  },
  "duration": 3612556,
  "status": "FAIL"
}

Executed 10 out of 13 tests successfully.
```
<a name="f6252ae1-2615-4aa0-85ea-b191b605c7f7"></a>
### 6419-b657ee0/mint-tls.sh.log:
```
Running with
SERVER_ENDPOINT:      localhost
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         1
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b7917b82200a:/mint/log /tmp/mint-logs'
(1/13) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-vcn11j3t3u3wyz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 317,
  "error": "RequestError: send request failed\ncaused by: Put https://localhost/aws-sdk-go-test-vcn11j3t3u3wyz: dial tcp 127.0.0.1:443: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 13 tests successfully.
```
<a name="3169f83b-70ed-4fb3-aa72-100321d00e18"></a>
### 6419-b657ee0/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31155
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3cb8e1612573:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 54 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 1 seconds
(4/13) Running aws-sdk-php tests ... done in 43 seconds
(5/13) Running aws-sdk-ruby tests ... done in 8 seconds
(6/13) Running mc tests ... done in 18 seconds
(7/13) Running minio-dotnet tests ... done in 3 minutes and 40 seconds
(8/13) Running minio-go tests ... done in 48 seconds
(9/13) Running minio-java tests ... done in 1 minutes and 39 seconds
(10/13) Running minio-js tests ... done in 36 seconds
(11/13) Running minio-py tests ... FAILED in 1 hours 4 minutes and 24 seconds
{
  "duration": 3612382,
  "function": "put_object(bucket_name, object_name, data, length, content_type, metadata)",
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 600, in urlopen\n    chunked=chunked)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 384, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 2, in raise_from\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 380, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.5/http/client.py\", line 1197, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.5/http/client.py\", line 297, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.5/http/client.py\", line 266, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1427, in test_thread_safe\n    client.remove_object(bucket_name, object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 942, in remove_object\n    object_name=object_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1752, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 322, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 638, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 398, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minikube', port=31155): Max retries exceeded with url: /minio-py-test-32ddede4-6c12-4993-8153-41f8fc9da5c1/d9178225-f60e-4c95-88c3-e63a2df54c3a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1800, in main\n    test_thread_safe(client, testfile, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1430, in test_thread_safe\n    raise Exception(err)\nException: HTTPConnectionPool(host='minikube', port=31155): Max retries exceeded with url: /minio-py-test-32ddede4-6c12-4993-8153-41f8fc9da5c1/d9178225-f60e-4c95-88c3-e63a2df54c3a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n",
  "message": "HTTPConnectionPool(host='minikube', port=31155): Max retries exceeded with url: /minio-py-test-32ddede4-6c12-4993-8153-41f8fc9da5c1/d9178225-f60e-4c95-88c3-e63a2df54c3a (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))",
  "args": {
    "object_name": "d9178225-f60e-4c95-88c3-e63a2df54c3a",
    "bucket_name": "minio-py-test-32ddede4-6c12-4993-8153-41f8fc9da5c1"
  }
}

Executed 10 out of 13 tests successfully.
```
<a name="c4ffe113-340a-45e8-ba81-0f199ecacb4d"></a>
### 6419-b657ee0/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32201
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 531aa7b9fbfb:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 49 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 1 seconds
(4/13) Running aws-sdk-php tests ... done in 40 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... done in 16 seconds
(7/13) Running minio-dotnet tests ... done in 1 minutes and 22 seconds
(8/13) Running minio-go tests ... done in 27 seconds
(9/13) Running minio-java tests ... done in 1 minutes and 19 seconds
(10/13) Running minio-js tests ... done in 35 seconds
(11/13) Running minio-py tests ... FAILED in 30 minutes and 6 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 600, in urlopen\n    chunked=chunked)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 384, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 2, in raise_from\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 380, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.5/http/client.py\", line 1197, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.5/http/client.py\", line 297, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.5/http/client.py\", line 266, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 165, in test_make_bucket_default_region\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 286, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 322, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 638, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 398, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minikube', port=32201): Max retries exceeded with url: /minio-py-test-1da8fbe9-c51d-42a2-aa7c-1c9221549540/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1707, in main\n    test_make_bucket_default_region(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 173, in test_make_bucket_default_region\n    raise Exception(err)\nException: HTTPConnectionPool(host='minikube', port=32201): Max retries exceeded with url: /minio-py-test-1da8fbe9-c51d-42a2-aa7c-1c9221549540/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n",
  "function": "make_bucket(bucket_name, location)",
  "args": {
    "location": "default value ('us-east-1')",
    "bucket_name": "minio-py-test-1da8fbe9-c51d-42a2-aa7c-1c9221549540"
  },
  "status": "FAIL",
  "name": "minio-py:test_make_bucket_default_region",
  "duration": 1806022,
  "message": "HTTPConnectionPool(host='minikube', port=32201): Max retries exceeded with url: /minio-py-test-1da8fbe9-c51d-42a2-aa7c-1c9221549540/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))"
}

Executed 10 out of 13 tests successfully.
```

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -178,6 +178,10 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
 		writeErrorResponse(w, ErrInvalidQuoteFields, r.URL)
 		return
 	}
+	if len(selectReq.InputSerialization.CSV.RecordDelimiter) > 2 {
+		writeErrorResponse(w, ErrInvalidRequestParameter, r.URL)
+		return
+	}
 
 	getObject := objectAPI.GetObject
 	if api.CacheAPI() != nil && !crypto.SSEC.IsRequested(r.Header) {
@@ -222,9 +226,13 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
 	if selectReq.InputSerialization.CSV.FileHeaderInfo == "" {
 		selectReq.InputSerialization.CSV.FileHeaderInfo = CSVFileHeaderInfoNone
 	}
+	if selectReq.InputSerialization.CSV.RecordDelimiter == "" {
+		selectReq.InputSerialization.CSV.RecordDelimiter = "\n"
+	}
 	if selectReq.InputSerialization.CSV != nil {
 		options := &s3select.Options{
 			HasHeader:            selectReq.InputSerialization.CSV.FileHeaderInfo != CSVFileHeaderInfoNone,
+			RecordDelimiter:      selectReq.InputSerialization.CSV.RecordDelimiter,
 			FieldDelimiter:       selectReq.InputSerialization.CSV.FieldDelimiter,
 			Comments:             selectReq.InputSerialization.CSV.Comments,
 			Name:                 "S3Object", // Default table name for all objects

--- a/pkg/ioutil/delimited-reader.go
+++ b/pkg/ioutil/delimited-reader.go
@@ -1,0 +1,87 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ioutil
+
+import (
+	"bufio"
+	"io"
+)
+
+var (
+	nByte byte = 10 // the byte that corresponds to the '\n' rune.
+	rByte byte = 13 // the byte that corresponds to the '\r' rune.
+)
+
+// DelimitedReader reduces the custom delimiter to `\n`.
+type DelimitedReader struct {
+	r           *bufio.Reader
+	delimiter   []rune // Select can have upto 2 characters as delimiter.
+	assignEmpty bool   // Decides whether the next read byte should be discarded.
+}
+
+// NewDelimitedReader detects the custom delimiter and replaces with `\n`.
+func NewDelimitedReader(r io.Reader, delimiter []rune) *DelimitedReader {
+	return &DelimitedReader{r: bufio.NewReader(r), delimiter: delimiter, assignEmpty: false}
+}
+
+// Reads and replaces the custom delimiter with `\n`.
+func (r *DelimitedReader) Read(p []byte) (n int, err error) {
+	n, err = r.r.Read(p)
+	if err != nil {
+		return
+	}
+	for i, b := range p {
+		if r.assignEmpty {
+			swapAndNullify(p, i)
+			r.assignEmpty = false
+			continue
+		}
+		if b == rByte && rune(b) != r.delimiter[0] {
+			// Replace the carriage returns with `\n`.
+			// Mac styled csv will have `\r` as their record delimiter.
+			p[i] = nByte
+		} else if rune(b) == r.delimiter[0] { // Eg, `\r\n`,`ab`,`a` are valid delimiters
+			if i+1 == len(p) && len(r.delimiter) > 1 {
+				// If the first delimiter match falls on the boundary,
+				// Peek the next byte and if it matches, discard it in the next byte read.
+				if nextByte, nerr := r.r.Peek(1); nerr == nil {
+					if rune(nextByte[0]) == r.delimiter[1] {
+						p[i] = nByte
+						// To Discard in the next read.
+						r.assignEmpty = true
+					}
+				}
+			} else if len(r.delimiter) > 1 && rune(p[i+1]) == r.delimiter[1] {
+				// The second delimiter falls in the same chunk.
+				p[i] = nByte
+				r.assignEmpty = true
+			} else if len(r.delimiter) == 1 {
+				// Replace with `\n` incase of single charecter delimiter match.
+				p[i] = nByte
+			}
+		}
+	}
+	return
+}
+
+// Occupy the first byte space and nullify the last byte.
+func swapAndNullify(p []byte, n int) {
+	for i := n; i < len(p)-1; i++ {
+		p[i] = p[i+1]
+	}
+	p[len(p)-1] = 0
+}

--- a/pkg/ioutil/delimited-reader_test.go
+++ b/pkg/ioutil/delimited-reader_test.go
@@ -1,0 +1,83 @@
+/*
+ * Minio Cloud Storage, (C) 2016, 2017, 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ioutil
+
+import (
+	"bytes"
+	"io"
+	"strings"
+	"testing"
+)
+
+// Test for DelimitedCSVReader.
+func TestDelimitedReader(t *testing.T) {
+	expected := "username,age\nbanana,12\ncarrot,23\napple,34\nbrinjal,90\nraddish,45"
+
+	inputs := []struct {
+		inputcsv  string
+		delimiter string
+		chunkSize int
+	}{
+		// case 1 - with default `\n` delimiter.
+		{"username,age\nbanana,12\ncarrot,23\napple,34\nbrinjal,90\nraddish,45", "\n", 10},
+		// case 2 - with carriage return `\r` which should be replaced with `\n` by default.
+		{"username,age\rbanana,12\rcarrot,23\rapple,34\rbrinjal,90\rraddish,45", "\n", 10},
+		// case 3 - with a double character delimiter (octals).
+		{"username,age\r\nbanana,12\r\ncarrot,23\r\napple,34\r\nbrinjal,90\r\nraddish,45", "\r\n", 10},
+		// case 4 - with a double character delimiter.
+		{"username,agexvbanana,12xvcarrot,23xvapple,34xvbrinjal,90xvraddish,45", "xv", 10},
+		// case 5 - with a double character delimiter `\t `
+		{"username,age\t banana,12\t carrot,23\t apple,34\t brinjal,90\t raddish,45", "\t ", 10},
+		// case 6 - This is a special case where the first delimiter match falls in the 13'th byte space
+		// ie, the last byte space of the read chunk, In this case the reader should peek in the next byte
+		// and replace with `\n`.
+		{"username,agexxbanana,12xxcarrot,23xxapple,34xxbrinjal,90xxraddish,45", "xx", 13},
+	}
+
+	for c, input := range inputs {
+		var readcsv []byte
+		var err error
+		delimitedReader := NewDelimitedReader(strings.NewReader(input.inputcsv), []rune(input.delimiter))
+		for err == nil {
+			chunk := make([]byte, input.chunkSize)
+			_, err = delimitedReader.Read(chunk)
+			readcsv = append(readcsv, chunk...)
+		}
+		if err != io.EOF {
+			t.Fatalf("Case %d: Error in delimited read", c+1)
+		}
+		expected := []byte(expected)
+		cleanCsv := removeNulls(readcsv)
+		if !bytes.Equal(cleanCsv, expected) {
+			t.Fatalf("Case %d: Expected the delimited csv to be `%s`, but instead found `%s`", c+1, string(expected), string(cleanCsv))
+		}
+	}
+
+}
+
+// Removes all the tailing nulls in chunks.
+// Null chunks will be assigned if there is a reduction
+// Eg, When `xv` is reduced to `\n`, the last byte is nullified.
+func removeNulls(csv []byte) []byte {
+	cleanCsv := []byte{}
+	for _, p := range csv {
+		if p != 0 {
+			cleanCsv = append(cleanCsv, p)
+		}
+	}
+	return cleanCsv
+}

--- a/pkg/s3select/input.go
+++ b/pkg/s3select/input.go
@@ -29,6 +29,7 @@ import (
 	"net/http"
 
 	gzip "github.com/klauspost/pgzip"
+	"github.com/minio/minio/pkg/ioutil"
 )
 
 const (
@@ -79,6 +80,9 @@ type Options struct {
 	// HasHeader when true, will treat the first row as a header row.
 	HasHeader bool
 
+	// RecordDelimiter is the string that records are delimited by.
+	RecordDelimiter string
+
 	// FieldDelimiter is the string that fields are delimited by.
 	FieldDelimiter string
 
@@ -127,7 +131,8 @@ func NewInput(opts *Options) (*Input, error) {
 		tempBytesScanned = opts.StreamSize
 		myReader = bzip2.NewReader(opts.ReadFrom)
 	}
-
+	// DelimitedReader treats custom record delimiter like `\r\n`,`\r`,`ab` etc and replaces it with `\n`.
+	normalizedReader := ioutil.NewDelimitedReader(myReader, []rune(opts.RecordDelimiter))
 	progress := &statInfo{
 		BytesScanned:   tempBytesScanned,
 		BytesProcessed: 0,
@@ -135,7 +140,7 @@ func NewInput(opts *Options) (*Input, error) {
 	}
 	reader := &Input{
 		options: opts,
-		reader:  csv.NewReader(myReader),
+		reader:  csv.NewReader(normalizedReader),
 		stats:   progress,
 	}
 	reader.firstRow = nil

--- a/pkg/s3select/select_test.go
+++ b/pkg/s3select/select_test.go
@@ -48,6 +48,7 @@ func TestCheckForDuplicates(t *testing.T) {
 func TestMyProcessing(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -96,6 +97,7 @@ func TestMyProcessing(t *testing.T) {
 func TestMyRowIndexResults(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -271,6 +273,7 @@ func TestMyParser(t *testing.T) {
 	for _, table := range tables {
 		options := &Options{
 			HasHeader:            false,
+			RecordDelimiter:      "\n",
 			FieldDelimiter:       ",",
 			Comments:             "",
 			Name:                 "S3Object", // Default table name for all objects
@@ -346,6 +349,7 @@ func TestMyAggregationFunc(t *testing.T) {
 func TestToStringAgg(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -380,6 +384,7 @@ func TestToStringAgg(t *testing.T) {
 func TestMyRowColLiteralResults(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -459,6 +464,7 @@ func TestMyWhereEval(t *testing.T) {
 	for _, table := range tables {
 		options := &Options{
 			HasHeader:            false,
+			RecordDelimiter:      "\n",
 			FieldDelimiter:       ",",
 			Comments:             "",
 			Name:                 "S3Object", // Default table name for all objects
@@ -610,6 +616,7 @@ func TestInterpreter(t *testing.T) {
 	for _, table := range tables {
 		options := &Options{
 			HasHeader:            false,
+			RecordDelimiter:      "\n",
 			FieldDelimiter:       ",",
 			Comments:             "",
 			Name:                 "S3Object", // Default table name for all objects
@@ -651,6 +658,7 @@ func TestInterpreter(t *testing.T) {
 func TestMyXMLFunction(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -689,6 +697,7 @@ func TestMyXMLFunction(t *testing.T) {
 func TestMyProtocolFunction(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -732,6 +741,7 @@ func TestMyProtocolFunction(t *testing.T) {
 func TestMyInfoProtocolFunctions(t *testing.T) {
 	options := &Options{
 		HasHeader:            true,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -773,6 +783,7 @@ func TestMyInfoProtocolFunctions(t *testing.T) {
 func TestMyErrorProtocolFunctions(t *testing.T) {
 	options := &Options{
 		HasHeader:            false,
+		RecordDelimiter:      "\n",
 		FieldDelimiter:       ",",
 		Comments:             "",
 		Name:                 "S3Object", // Default table name for all objects
@@ -1007,6 +1018,7 @@ func TestMyValids(t *testing.T) {
 	for _, table := range tables {
 		options := &Options{
 			HasHeader:            false,
+			RecordDelimiter:      "\n",
 			FieldDelimiter:       ",",
 			Comments:             "",
 			Name:                 "S3Object", // Default table name for all objects

